src/sentry/api/endpoints/project_metrics_extraction_rules.py

```diff
@@ -81,62 +83,90 @@ def post(self, request: Request, project: Project) -> Response:
-                persisted_rules, request.user, MetricsExtractionRuleSerializer()
-            )
-        except Exception as e:
-            return Response(status=400, data={"detail": str(e)})
-
-        return Response(status=200, data=updated_rules)
-
-    def _generate_updated_rule_objects(
-        self, updated_rules: list[dict[str, Any]]
-    ) -> dict[str, MetricsExtractionRule]:
-        state_update = {}
-        for updated_rule in updated_rules:
-            rule = MetricsExtractionRule.from_dict(updated_rule)
-            mri = rule.generate_mri()
-            state_update[mri] = rule
-
-        return state_update
-
-    def _generate_deleted_rule_objects(
-        self, updated_rules: list[dict[str, Any]]
-    ) -> Sequence[MetricsExtractionRule]:
-        state_update = []
-        for updated_rule in updated_rules:
-            state_update.append(MetricsExtractionRule.from_dict(updated_rule))
+            with transaction.atomic(router.db_for_write(SpanAttributeExtractionRuleConfig)):
+                configs = []
+                for obj in config_update:
+                    config = SpanAttributeExtractionRuleConfig.objects.get(
+                        project=project, span_attribute=obj["spanAttribute"]
+                    )
+                    config.aggregates = obj["aggregates"]
+                    config.unit = HARD_CODED_UNITS.get(obj["spanAttribute"], obj["unit"])
+                    config.tags = obj["tags"]
+                    config.save()
+                    config.refresh_from_db()
+                    configs.append(config)
+
+                    # delete conditions not present in update
+                    included_conditions = [x["id"] for x in obj["conditions"]]
+                    SpanAttributeExtractionRuleCondition.objects.filter(
+                        project=project, config=config
+                    ).exclude(id__in=included_conditions).delete()
+
+                    for condition in obj["conditions"]:
+                        condition_exists = SpanAttributeExtractionRuleCondition.objects.filter(
+                            id=condition["id"], project=project
+                        ).exists()
+
+                        # update existing conditions
+                        if condition_exists:
+                            SpanAttributeExtractionRuleCondition.objects.filter(
+                                id=condition["id"], project=project
+                            ).update(value=condition["value"])
+
+                        # create new conditions
+                        else:
+                            condition_obj = SpanAttributeExtractionRuleCondition.objects.create(
+                                created_by_id=request.user.id,
+                                project=project,
+                                value=condition["value"],
+                                config=config,
+                            )
+                            condition_obj.save()
```

Comment:
More Django way of doing this would be to use `update_or_create` ([docs](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#update-or-create)):

```suggestion
                        condition_id = condition["id'] if 'id' in condition else None
                        SpanAttributeExtractionRuleCondition.objects.update_or_create(id=condition_id, project=project, config=config, defaults={"value":condition["value"], "created_by_id":request.user.id}
```
Django will first check if the object exist and all the values in `default` dict will be updated. If it doesn't exist, then new record will be created with all the values passed in args

Reply:
Nice suggestion, changed